dimod/bqm/adjdictbqm.py

```diff
@@ -652,3 +652,89 @@ def set_quadratic(self, u, v, bias):
 
         adj = self._adj
         adj[u][v] = adj[v][u] = bias
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def __mul__(self, other):
+        if isinstance(other, (int, float)):
+            bqm = AdjDictBQM(vartype=self.vartype)
+            bqm.add_offset(self.offset * other)
+            for v, bias in self.linear.items():
+                bqm.add_variable(v, bias * other)
+
+            for (u, v), bias in self.quadratic.items():
+                bqm.add_interaction(u, v, bias * other)
+
+        elif isinstance(other, BQM):
+            if other.shape[1] > 0 or self.shape[1] > 0:
+                raise ValueError("Multiply is supported for BQMs with "
+                                 "linear terms only")
+            if other.vartype != self.vartype:
+                other = other.change_vartype(self.vartype)
+
+            bqm = AdjDictBQM(vartype=self.vartype)
+            bqm.add_offset(self.offset * other.offset)
+            for v, bias in other.linear.items():
+                bqm.add_variable(v, bias * self.offset)
+            for v, bias in self.linear.items():
+                bqm.add_variable(v, bias * other.offset)
+
+            for v, bias in self.linear.items():
+                for u, bias_u in other.linear.items():
+                    if v == u:
+                        if bqm.vartype == Vartype.BINARY:
+                            bqm.add_variable(v, bias * bias_u)
+                        else:
+                            bqm.add_offset(bias * bias_u)
+                    else:
+                        bqm.add_interaction(u, v, bias_u * bias)
+        else:
+            raise ValueError("other must be a numeric or an instance of BQM")
+        return bqm
+
+    def __sub__(self, other):
+        return self.__add__(-other)
+
+    def __neg__(self):
+        return -1 * self
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __add__(self, other):
+
+        if isinstance(other, (int, float)):
+            bqm = self.copy()
+            bqm.add_offset(other)
+
+        elif isinstance(other, BQM):
+            if other.vartype != self.vartype:
+                other = other.change_vartype(self.vartype)
+            bqm = self.copy()
+            bqm.add_offset(other.offset)
+            for v, bias in other.linear.items():
+                bqm.add_variable(v, bias)
+
+            for (u, v), bias in other.quadratic.items():
+                bqm.add_interaction(u, v, bias)
+        else:
+            raise ValueError("other must be a numeric or an instance of BQM")
+        return bqm
+
+    def __pow__(self, power, modulo=None):
```

Comment:
One question is, should we allow this at all or just have the user do `bqm * bqm`, which is more explicit.